include/kernels/coarse_op_kernel_mma.cuh

```diff
@@ -52,60 +74,84 @@ namespace quda
         constexpr int ldb = N;
         constexpr int ldc = N;
 
-        using Config = MmaConfig<M, N, K, lda, ldb, ldc, bM, bN, bK, block_y, block_z>;
+        using Config = MmaConfig<M, N, K, lda, ldb, ldc, Arg::bM, Arg::bN, Arg::bK, Arg::block_y, Arg::block_z>;
+
+        if (Arg::dir == QUDA_IN_PLACE) {
 
-        if (arg.comm_dim[dim] && (coord[dim] + nFace >= arg.x_size[dim])) {
+          for (int s_col = 0; s_col < fineSpin; s_col++) {
 
-          int ghost_idx = ghostFaceIndex<1>(coord, arg.x_size, dim, nFace);
+            auto a = arg.C(0, parity, x_cb, 0, s_col, 0, 0);
+            auto b = Wacc(parity, x_cb, s_col, 0, 0);
+            auto c = arg.UV(parity, x_cb, s_col * fineSpin, 0, 0);
+
+            uvMax = Config::template perform_mma<a_dagger, b_dagger, Arg::compute_max>(a, b, c, 0, 0);
+          }
+
+        } else if (arg.comm_dim[Arg::dim] && (coord[Arg::dim] + nFace >= arg.x_size[Arg::dim])) {
+
+          int ghost_idx = ghostFaceIndex<1>(coord, arg.x_size, Arg::dim, nFace);
 
           for (int s_col = 0; s_col < fineSpin; s_col++) {
 
-            auto a = arg.U.wrap(dim + (dir == QUDA_FORWARDS ? 4 : 0), parity, x_cb, 0, s_col);
-            auto b = Wacc.wrap_ghost(dim, 1, (parity + 1) & 1, ghost_idx, s_col);
-            auto c = arg.UV.wrap(parity, x_cb, s_col * fineSpin);
+            auto a = arg.U(Arg::dim + (Arg::dir == QUDA_FORWARDS ? 4 : 0), parity, x_cb, 0, s_col, 0, 0);
+            auto b = Wacc.Ghost(Arg::dim, 1, (parity + 1) & 1, ghost_idx, s_col, 0, 0);
+            auto c = arg.UV(parity, x_cb, s_col * fineSpin, 0, 0);
 
-            Config::template perform_mma<a_dagger, b_dagger, compute_max_only>(a, b, c, 0, 0);
+            uvMax = Config::template perform_mma<a_dagger, b_dagger, Arg::compute_max>(a, b, c, 0, 0);
           }
 
         } else {
 
-          int y_cb = linkIndexP1(coord, arg.x_size, dim);
+          int y_cb = linkIndexP1(coord, arg.x_size, Arg::dim);
 
           for (int s_col = 0; s_col < fineSpin; s_col++) {
 
-            auto a = arg.U.wrap(dim + (dir == QUDA_FORWARDS ? 4 : 0), parity, x_cb, 0, s_col);
-            auto b = Wacc.wrap((parity + 1) & 1, y_cb, s_col);
-            auto c = arg.UV.wrap(parity, x_cb, s_col * fineSpin);
+            auto a = arg.U(Arg::dim + (Arg::dir == QUDA_FORWARDS ? 4 : 0), parity, x_cb, 0, s_col, 0, 0);
+            auto b = Wacc((parity + 1) & 1, y_cb, s_col, 0, 0);
+            auto c = arg.UV(parity, x_cb, s_col * fineSpin, 0, 0);
 
-            Config::template perform_mma<a_dagger, b_dagger, compute_max_only>(a, b, c, 0, 0);
+            uvMax = Config::template perform_mma<a_dagger, b_dagger, Arg::compute_max>(a, b, c, 0, 0);
           }
         }
+
+        return uvMax;
       } // computeUV
 
     } // namespace impl
 
-    template <bool from_coarse, int dim, QudaDirection dir, int bM, int bN, int bK, int block_y, int block_z, typename Arg>
-    __global__ void __launch_bounds__(block_y *block_z, 1) ComputeUVMMA(Arg arg)
```

Comment:
I wonder where does `__launch_bounds__` go in this case?

Reply:
For the tensor core kernels I made a generic `raw_kernel` that does not try to abstract the parallelism, and leaves this all to the functor.  This `raw_kernel` expects that the `Arg` class has `static constexpr int block_dim` and `static constexpr int min_blocks` members.

https://github.com/lattice/quda/blob/2e720eff0fd417a3967178bdf23247cf88101437/include/targets/cuda/kernel.h#L74